openshift/tests-extension/test/env/cluster.go

```diff
@@ -0,0 +1,140 @@
+// discovering whether the cluster is OpenShift, registering required API schemes,
+// and creating a controller-runtime client. This is used to build the shared TestEnv object
+// that provides access to the API and client in tests.
+// You should call this function before any suite that requires access to the cluster:
+// Example usage:
+//
+//		BeforeEach(func() {
+//	    _ = env.Init()
+//		})
+func initTestEnv() *TestEnv {
+	cfg := getRestConfig()
+
+	discoveryClient := discovery.NewDiscoveryClientForConfigOrDie(cfg)
+	isOcp := detectOpenShift(discoveryClient)
+
+	scheme := runtime.NewScheme()
+	utilruntime.Must(apiextensionsv1.AddToScheme(scheme))
+	if isOcp {
+		utilruntime.Must(configv1.AddToScheme(scheme))
+	}
+
+	k8sClient, err := crclient.New(cfg, crclient.Options{Scheme: scheme})
+	if err != nil {
+		log.Fatalf("failed to create controller-runtime client: %v", err)
+	}
+
+	if isOcp {
+		extlogs.Infof("[env] Cluster environment initialized (OpenShift: %t)", isOcp)
+	}
+
+	return &TestEnv{
+		RestCfg:     cfg,
+		K8sClient:   k8sClient,
+		IsOpenShift: isOcp,
+	}
+}
+
+// getRestConfig returns a Kubernetes REST config for the test client.
+// It first checks the KUBECONFIG environment variable and uses that if available.
+// If not, it falls back to using in-cluster configuration (when running inside a pod).
+// This allows the same test code to run in both local and cluster environments.
+func getRestConfig() *rest.Config {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	if _, err := os.Stat(kubeconfig); err == nil {
+		cfg, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			log.Fatalf("Failed to load kubeconfig from %s: %v", kubeconfig, err)
+		}
+		extlogs.Infof("[env] Using kubeconfig: %s", kubeconfig)
+		return configureQPS(cfg)
+	}
+	extlogs.Infof("[env] Using in-cluster configuration: %s", kubeconfig)
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		log.Fatalf("Failed to load in-cluster config: %v", err)
+	}
+	return configureQPS(cfg)
+}
+
+// detectOpenShift checks if the cluster is an OpenShift cluster.
+// It does this by looking for the "config.openshift.io" API group,
+// which only exists in OpenShift environments.
+func detectOpenShift(d discovery.DiscoveryInterface) bool {
+	groups, err := d.ServerGroups()
+	if err != nil {
+		extlogs.WarnContextf("failed to list API groups: %v", err)
+		return false
+	}
+	for _, g := range groups.Groups {
+		if g.Name == "config.openshift.io" {
+			return true
+		}
+	}
+	return false
+}
+
+// configureQPS sets high QPS and burst values to avoid client-side throttling during tests.
+// This makes tests faster by allowing many API calls without delay.
+// It's mainly needed in serial tests, where slow or throttled requests can cause flakes.
+// The default limits (QPS=5, Burst=10) are too low for most test workloads.
+func configureQPS(cfg *rest.Config) *rest.Config {
+	cfg.QPS = 10000
+	cfg.Burst = 10000
+	cfg.RateLimiter = nil
```

Comment:
See that ocp/origin client utils will create a TOM of complexities to create a fakeRateLimiter that return nill at the end : https://github.com/openshift/kubernetes/blob/b94367cabf3edea99cfd7ecf7e8bafdc908d83ca/staging/src/k8s.io/client-go/util/flowcontrol/throttle.go#L152-L154